src/resolve/resolved-dns-query.c

```diff
@@ -312,6 +316,22 @@ static int dns_query_candidate_setup_transactions(DnsQueryCandidate *c) {
         return r;
 }
 
+static void dns_query_accept(DnsQuery *q, DnsQueryCandidate *c);
+
+static int on_candidate_timeout(sd_event_source *s, usec_t usec, void *userdata) {
+        DnsQueryCandidate *c = userdata;
+
+        assert(s);
+        assert(c);
+
+        if (DNS_TRANSACTION_IS_LIVE(c->query->state)) {
```

Comment:
we generally prefer "early exit" over "deep nesting". hence please invert this check and exit the function.

Can this condition actually ever not happen? not really? maybe treat this as an assert? or at least debug log?

Reply:
Looked through the code again. As far as I can tell, the only way for the transaction to become not-live is through `dns_query_complete`, which calls `dns_query_abandon` which calls `dns_query_candidate_abandon`.
I've removed this check, and added a missing `event_source_disable(c->timeout_event_source)` to `dns_query_candidate_abandon`